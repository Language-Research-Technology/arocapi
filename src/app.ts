import cors from '@fastify/cors';
import sensible from '@fastify/sensible';
import type { Client } from '@opensearch-project/opensearch';
import type { PrismaClient } from '@prisma/client/extension';
import type { FastifyInstance, FastifyPluginAsync } from 'fastify';
import fp from 'fastify-plugin';
import {
  hasZodFastifySchemaValidationErrors,
  isResponseSerializationError,
  serializerCompiler,
  validatorCompiler,
} from 'fastify-type-provider-zod';
import entities from './routes/entities.js';
import entity from './routes/entity.js';
import search from './routes/search.js';

const setupValidation = (fastify: FastifyInstance) => {
  fastify.setValidatorCompiler(validatorCompiler);
  fastify.setSerializerCompiler(serializerCompiler);

  fastify.setErrorHandler((err, req, reply) => {
    if (hasZodFastifySchemaValidationErrors(err)) {
      return reply.code(400).send({
        error: 'Response Validation Error',
        message: "Request doesn't match the schema",
        details: {
          issues: err.validation,
          method: req.method,
          url: req.url,
        },
      });
    }

    if (isResponseSerializationError(err)) {
      return reply.code(500).send({
        error: 'Internal Server Error',
        message: "Response doesn't match the schema",
        details: {
          issues: err.cause.issues,
          method: err.method,
          url: err.url,
        },
      });
    }

    // NOTE: We are exposing the error message here for development purposes.
    // In production, consider hiding error details to avoid leaking sensitive information.
    return reply.code(500).send({
      error: 'Internal Server Error',
      message: err.message,
    });
  });
};

const setupDatabase = async (fastify: FastifyInstance, prisma: PrismaClient) => {
  await prisma.$connect();
  fastify.decorate('prisma', prisma);
  fastify.addHook('onClose', async (server) => {
    await server.prisma.$disconnect();
  });
};

const setupSearch = async (fastify: FastifyInstance, opensearch: Client) => {
  try {
    await opensearch.ping();
    fastify.log.info(`Connected to OpenSearch`);
  } catch (error) {
    fastify.log.error(`Failed to connect to OpenSearch: ${error}`);
    throw error;
  }

  fastify.decorate('opensearch', opensearch);

  fastify.addHook('onClose', () => opensearch.close());
};

export type Options = {
  prisma: PrismaClient;
  opensearch: Client;
  disableCors?: boolean;
};
const app: FastifyPluginAsync<Options> = async (fastify, options) => {
  const { prisma, opensearch, disableCors = false } = options || {};

  fastify.register(sensible);
  if (!disableCors) {
    fastify.register(cors);
  }
  setupValidation(fastify);
  await setupDatabase(fastify, prisma);
  await setupSearch(fastify, opensearch);

  fastify.register(entities);
  fastify.register(entity);
  fastify.register(search);
};

export default fp(app);
