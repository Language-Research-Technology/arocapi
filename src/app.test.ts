import Fastify from 'fastify';
import { beforeEach, describe, expect, it } from 'vitest';
import { mockReset } from 'vitest-mock-extended';

import app from './app.js';

import { opensearch, prisma } from './test/helpers/fastify.js';

describe('Entity Route', () => {
  beforeEach(() => {
    mockReset(prisma);
    mockReset(opensearch);
  });

  describe('App Registration', () => {
    it('should handle missing prisma', async () => {
      const fastify = Fastify({ logger: false });

      // @ts-expect-error we are testing missing options
      fastify.register(app, {
        opensearch,
        disableCors: false,
      });

      await expect(() => fastify.ready()).rejects.toThrowError('Prisma client is required');
    });

    it('should handle missing opensearch', async () => {
      const fastify = Fastify({ logger: false });

      // @ts-expect-error we are testing missing options
      fastify.register(app, {
        prisma,
        disableCors: false,
      });

      await expect(() => fastify.ready()).rejects.toThrowError('OpenSearch client is required');
    });

    it('should handle broken opensearch', async () => {
      opensearch.ping.mockRejectedValue(new Error('Connection failed'));

      const fastify = Fastify({ logger: false });

      fastify.register(app, {
        prisma,
        opensearch,
        disableCors: false,
      });

      await expect(() => fastify.ready()).rejects.toThrowError('Connection failed');
    });

    it('should handle random errors', async () => {
      const fastify = Fastify({ logger: false });
      await fastify.register(app, {
        prisma,
        opensearch,
        disableCors: false,
      });
      fastify.get('/error', async () => {
        throw new Error('Random');
      });

      await fastify.ready();

      const response = await fastify.inject({
        method: 'GET',
        url: '/error',
      });
      const body = JSON.parse(response.body);

      expect(response.statusCode).toBe(500);
      expect(body).toMatchSnapshot();
    });
  });
});
