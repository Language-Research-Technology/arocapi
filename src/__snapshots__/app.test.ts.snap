// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Entity Route > App Registration > should handle random errors 1`] = `
{
  "error": "Internal Server Error",
  "message": "Random",
}
`;
